File: Python/Problems/Problems/problem3128.py
from .problem import Problem
import math


class Problem3128(Problem):
    def test(self):
        grid = [[0, 1, 0], [0, 1, 1], [0, 1, 0]]

        expected = 2

        result = self.numberOfRightTriangles(grid)

        print(result)

        return expected == result

    def numberOfRightTriangles(self, grid):
        n = len(grid)
        m = len(grid[0])
        rows = [0 for _ in range(n)]
        cols = [0 for _ in range(m)]

        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    rows[i] += 1
                    cols[j] +=1

        result = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] != 0 and rows[i] > 1 and cols[j] > 1:
                    result += (rows[i] - 1) * (cols[j] - 1)


        return result

File: C++/Problems/Problems/problem_3128.cpp
#include "problem_3128.h"

#include <iostream>
#include <bits/ostream.tcc>

bool problem_3128::test() {
    std::vector<std::vector<int>> grid  {  { 0, 1, 0 },  { 0, 1, 1 },  { 0, 1, 0 } };

    auto expected = 2;

    auto result = numberOfRightTriangles(grid);

    std::cout << result << std::endl;

    return expected == result;
}

long long problem_3128::numberOfRightTriangles(std::vector<std::vector<int>> &grid) {
    int n = grid.size();
    int m = grid[0].size();
    std::vector<int> rows(n);
    std::vector<int> cols(m);

    for (int i = 0; i < n; ++i)
    {
        for (int j = m - 1; j >= 0; --j)
        {
            if (grid[i][j] == 1)
            {
                rows[i]++;
                cols[j]++;
            }
        }
    }

    long long result = 0;
    for (int i = 0; i < n; ++i)
    {
        for (int j = 0; j < m; ++j)
        {
            if (grid[i][j] != 0 && rows[i] > 1 && cols[j] > 1)
            {
                result += (rows[i] - 1) * (cols[j] - 1);
            }
        }
    }

    return result;
}
